### ISSUE TITLE
Focus management

### ISSUE THREAD

**Issue by wojciechczerniak** [handsontable/handsontable#8939]
Currently, Handsontable focus is basically the same as selection. It's marked by active cell and is part of Selection layer. Styled as one, with bold border around a single cell that we're navigating with. 
Although, we have to remember that selection is not the same as focus. Selection may follow focus in some circumstances (see below) but they are not the same. WAI best practices defines focus as "focus is a pointer, like a mouse pointer; it tracks the path of navigation. There is only one point of focus at any time and all operations take place at the point of focus." 
Since we want to navigate/focus header cells, this implementation has to be changed.

### Tracking focused element

According to WAI focus has to be stored and restored when navigating through focus tree (tab sequence elements). Keyboard navigation requirements says that we should be able to move in and out of the component. First focus on Handsontable should internally focus on default element of our choice, that will be cell (1,1) for backward compatibility and current UX. But when we navigate to other cell, move out of Handsontable and then back (<kbd>Tab</kbd>, then <kbd>Shift</kbd>+<kbd>Tab</kbd>) we have to restore previously focused element. Therefore we need new layer that will store and remember last focused element.

Focusing Handsontable with pointer event, moves focus to the pointer event target.

### ARIA helper

Since only main `DIV` element with `role="grid"` should be included in Tab Sequence as a focusable element we should take a closer look at `aria-activedescendant` attribute that marks internal element of composite component (like grid) that should be focused when container (grid) is activated. Attribute value for grid component is dynamic, it will change when focused cell changes and should be remembered as described in previous section. There are two other attributes `aria-owns` and `aria-controls` that work together with `aria-activedecendant`. ARIA attributes will be described in more details as a separate issue.

### Focus levels

With composite component, like grid, there are few layers of focus:

1. First at main component layer where we can choose between top-layer sub-components: grid, toolbar, sidebar, pagination bar etc. Obviously, currently we will skip that choice and by default we will focus grid (Walkontable). We can skip to point (2), but to be future proof, we should be aware of that layer.

2. Next, we have grid layer where we navigate and focus cells (and headers, special case of a cell). We can have only one cell focused at a time. Which is different from selection, where we mark more cells, but move / expand selection by navigating with a single focused cell.

3. One level deeper are editors, renderers, context menu and dropdown menu. Focus should be then managed by editor because we can't know all the requirements and construction of such a component. Our editors will not have to remember last focused state as they're short living elements and it is expected to clear their state after accepting / discarding changes.

Between (2) and (3) we have a layer of Handsontable editor, that may have its own editors.

### Visual design

As described in [Focus VS Selection and the Perception of Dual Focus](https://w3c.github.io/aria-practices/#kbd_focus_vs_selection) we have to avoid mixing those two. In current visual design, active cell IS focus. A single focused cell that indicated current navigation pointer. Since we're going to have other elements that may be focused, this is no longer true. Focus pointer may be moved to headers and therefore, active cell has to loose the focused distinction and style.

> From the developer's perspective, the difference is simple -- the focused element is the active element (document.activeElement). Selected elements are elements that have `aria-selected="true"`.

> With respect to focus and the selected state, the most important considerations for designers and developers are:
>
> - The visual focus indicator must always be visible.
> - The selected state must be visually distinct from the focus indicator.

Also, focus style should be customisable to fit page/app design. It's very important to have focus indication consistent between page/app elements. Otherwise it's very easy to confuse users.

The best way to do so would be `:focus` state, but currently impossible with `<textarea>` autofocus for IME and Edit toolbar menu option. Therefore we need alternative method: `.ht-focused`  class on the focused element and `.ht-selected` (current `.area`) for all selected elements. `.ht-active` (current `.highlight`) for active cell would be a nice addition to the set.

Design for focused headers:
![a11y-headers-selection](https://user-images.githubusercontent.com/4718998/137911035-4c3102c4-c2b3-4090-a7f4-b8a62a3588af.jpg)
Note that active cell does not change its appearances. Neither, when losing focus. Therefore with above CSS selectors developer should be able to easily achieve WAI compatibility.

### Selection follow focus

Current default behaviour is called "selection follow focus". Each time we move active cell focus, we select new cell and deselect previous selection. This makes non-contiguous selection impossible to implement with single keyboard shortcut.

One way to make it work: Focus should navigate independently and for selection a modifier key should be required. "When selection does not follow focus, the user changes which element is selected by pressing the Enter or Space key." Since that would break our look&feel we're not going to change it.

Another way is inspired by Excel: Pressing <kbd>Shift</kbd>+<kbd>F8</kbd> key we can change selection mode. Current selection is kept in the background and "new layer" of selection is created. We can then use standard navigation and shortcut to create a new selection and/or extend it. To create another contiguous selection we have to use <kbd>Shift</kbd>+<kbd>F8</kbd> key again, and repeat the cycle.

Second option is the way we should follow.

### Disabled cells

Component authors may include disabled elements in focus path or not. The only requirement is to do so consequently. IMO Handsontable has this already handled very well. Selection for disabled cells is required to copy&paste content. Focus will be required to properly read the cell content with different user agents and select cells.

### `colspan` and `rowspan`

Current implementation of MergedCells should be transparent for focus and we should be able to correctly managed `.ht-focused`, `.ht-active` classes. We're translating coordinates to the top-left corner of merged range.
Nested Headers may be more complicated for full compatibility. Already mentioned "headers as cells" would make it easier to implement. 

### To do

- [ ] Skip (1) top-level focus, we don't have. At Handsontable level Walkontable is the default and only focus element. 
- [ ] We have nothing to do with (3) editor-level ATM. Editor authors are responsible for focus management when editor is activated. We may have to adjust our editors, but that's out of scope for now.
- [ ] Implement negative coordinates for headers to track currently focused header cell.
- [ ] Implement headers as cell, with renderer and edit/focus mode so we may navigate elements that are rendered within.
- [ ] Alter visual design to add focus layer and distinguish it correctly from selection layer.
- [ ] Focus style should be customisable so end dev may achieve visual consistency across whole page or app. A documentation have to be written for focus management and styling.
- [ ] Ensure disabled cells are focusable and can be correctly navigated to and from.
- [ ] Do not use `tabindex` to manage focus in any way.

**Comment by krzysztofspilka** [comment]
@wojciechczerniak will certainly make use of this one, thanks!
